src/AutoMapper/Execution/TypeMapPlanBuilder.cs

```diff
@@ -562,44 +562,37 @@ public Expression MapExpression(TypePair typePair, Expression sourceParameter, P
-        {
+            TypePair typePair, Expression sourceParameter, Expression contextParameter, PropertyMap propertyMap = null, Expression destinationParameter = null)
+        {
             if(destinationParameter == null)
             {
                 destinationParameter = Default(typePair.DestinationType);
-            }
-            var typeMap = configurationProvider.ResolveTypeMap(typePair);
-            if(typeMap != null)
-            {
-                if(!typeMap.HasDerivedTypesToInclude())
-                {
-                    typeMap.Seal(typeMapRegistry, configurationProvider);
-                    if(typeMap.MapExpression != null)
-                    {
-                        return typeMap.MapExpression.ConvertReplaceParameters(sourceParameter, destinationParameter, contextParameter);
-                    }
-                    else
-                    {
-                        return ContextMap(typePair, sourceParameter, contextParameter, destinationParameter);
-                    }
-                }
-                else
-                {
-                    return ContextMap(typePair, sourceParameter, contextParameter, destinationParameter);
-                }
+            }
+            var typeMap = configurationProvider.ResolveTypeMap(typePair);
+            if(typeMap != null)
+            {
+                return GetMapperExpression(configurationProvider, typeMapRegistry, typeMap)?.ConvertReplaceParameters(sourceParameter, destinationParameter, contextParameter)
+                       ?? ContextMap(typePair, sourceParameter, contextParameter, destinationParameter);
             }
             var match = configurationProvider.GetMappers().FirstOrDefault(m => m.IsMatch(typePair));
             if(match != null)
             {
                 var mapperExpression = match.MapExpression(typeMapRegistry, configurationProvider, propertyMap, sourceParameter, destinationParameter, contextParameter);
                 return ToType(mapperExpression, typePair.DestinationType);
             }
-            return ContextMap(typePair, sourceParameter, contextParameter, destinationParameter);
+            return ContextMap(typePair, sourceParameter, contextParameter, destinationParameter);
+        }
+
+        private static readonly MethodInfo GetMapperExpressionMethod = typeof(IConfigurationProvider).GetDeclaredMethod("GetMapperExpression");
+        private static LambdaExpression GetMapperExpression(IConfigurationProvider configurationProvider, TypeMapRegistry typeMapRegistry, TypeMap typeMap)
+        {
+            typeMap.Seal(typeMapRegistry, configurationProvider);
+            return GetMapperExpressionMethod.MakeGenericMethod(typeMap.SourceType, typeMap.DestinationType).Invoke(configurationProvider, new object[] { typeMap.Types }) as LambdaExpression;

```

Comment:
GetMapperExpression could be non generic and receive the types as objects and then you don't need reflection.